File: py/stripe_sdk/features.py
# Stripe SDK feature factory

from stripe_sdk.feature.base_feature import StripeBaseFeature
from stripe_sdk.feature.debug_feature import StripeDebugFeature
from stripe_sdk.feature.idempotency_feature import StripeIdempotencyFeature
from stripe_sdk.feature.metrics_feature import StripeMetricsFeature
from stripe_sdk.feature.paging_feature import StripePagingFeature
from stripe_sdk.feature.ratelimit_feature import StripeRatelimitFeature
from stripe_sdk.feature.retry_feature import StripeRetryFeature
from stripe_sdk.feature.test_feature import StripeTestFeature
from stripe_sdk.feature.timeout_feature import StripeTimeoutFeature


_FEATURES = {
    "base": lambda: StripeBaseFeature(),
    "debug": lambda: StripeDebugFeature(),
    "idempotency": lambda: StripeIdempotencyFeature(),
    "metrics": lambda: StripeMetricsFeature(),
    "paging": lambda: StripePagingFeature(),
    "ratelimit": lambda: StripeRatelimitFeature(),
    "retry": lambda: StripeRetryFeature(),
    "test": lambda: StripeTestFeature(),
    "timeout": lambda: StripeTimeoutFeature(),
}


def _make_feature(name):
    factory = _FEATURES.get(name)
    if factory is not None:
        return factory()
    return _FEATURES["base"]()


# True when this SDK was generated with the named feature class - the
# constructor's tolerance for extend-carried features reads this (an
# active name with no generated class must not become a BaseFeature
# stray when an extend instance carries it).
def _has_feature(name):
    return name in _FEATURES

File: php/features.php
<?php
declare(strict_types=1);

// Stripe SDK feature factory

require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/feature/DebugFeature.php';
require_once __DIR__ . '/feature/IdempotencyFeature.php';
require_once __DIR__ . '/feature/MetricsFeature.php';
require_once __DIR__ . '/feature/PagingFeature.php';
require_once __DIR__ . '/feature/RatelimitFeature.php';
require_once __DIR__ . '/feature/RetryFeature.php';
require_once __DIR__ . '/feature/TestFeature.php';
require_once __DIR__ . '/feature/TimeoutFeature.php';


class StripeFeatures
{
    public static function make_feature(string $name)
    {
        switch ($name) {
            case "base":
                return new StripeBaseFeature();
            case "debug":
                return new StripeDebugFeature();
            case "idempotency":
                return new StripeIdempotencyFeature();
            case "metrics":
                return new StripeMetricsFeature();
            case "paging":
                return new StripePagingFeature();
            case "ratelimit":
                return new StripeRatelimitFeature();
            case "retry":
                return new StripeRetryFeature();
            case "test":
                return new StripeTestFeature();
            case "timeout":
                return new StripeTimeoutFeature();
            default:
                return new StripeBaseFeature();
        }
    }

    /**
     * Does a generated feature class back this name? False for a name only
     * an options extend instance can supply (the station adopt path) - the
     * constructor uses this to skip make_feature for such names instead of
     * adding a stray BaseFeature.
     */
    public static function has_feature(string $name): bool
    {
        switch ($name) {
            case "base":
            case "debug":
            case "idempotency":
            case "metrics":
            case "paging":
            case "ratelimit":
            case "retry":
            case "test":
            case "timeout":
                return true;
            default:
                return false;
        }
    }
}
